wrap.c

```diff
@@ -0,0 +1,212 @@
+    strcpy(*newenv + var_len + 1, var_contents);
+#ifdef __M2__
+    return newenv + sizeof(char *);
+#else
+    return newenv + 1;
+#endif
+  }
+  return newenv;
+}
+
+int main(int argc, char **argv) {
+  if(argc <= 1) {
+    fputs("Expected at least one argument: command\n", stderr);
+    exit(EXIT_FAILURE);
+  }
+  char *cwd = get_current_dir_name();
+  /* Do nothing if cwd is already root */
+  if (strcmp(cwd, "/")) {
+    int uid = geteuid();
+    int gid = getegid();
+    /* Don't create a user and mount namespace if we are already root */
+    if (uid != 0) {
+      if (unshare(CLONE_NEWUSER | CLONE_NEWNS) != 0) {
+        fputs("Failed to create user and mount namespaces\n", stderr);
+        exit(EXIT_FAILURE);
+      }
+      /* Prevent the use of setgroups and make gid_map writeable */
+      deny_setgroups();
+      /* Map the root user in the user namespace to our user id */
+      set_map(uid, "/proc/self/uid_map");
+      /* Map the root group in the user namespace to our group id */
+      set_map(gid, "/proc/self/gid_map");
+    }
+    int r = mkdir("dev", 0755);
+    if (r != 0 && r != -17) {
+      fputs("Failed to create dev folder\n", stderr);
+      exit(EXIT_FAILURE);
+    }
+#if !__uefi__
+    mkmount ("/dev/null", "dev/null", "", MS_BIND, NULL, 0);
+    mkmount ("/dev/zero", "dev/zero", "", MS_BIND, NULL, 0);
+    mkmount ("/dev/random", "dev/random", "", MS_BIND, NULL, 0);
+    mkmount ("/dev/urandom", "dev/urandom", "", MS_BIND, NULL, 0);
+    mkmount ("/dev/ptmx", "dev/ptmx", "", MS_BIND, NULL, 0);
+    mkmount ("/dev/tty", "dev/tty", "", MS_BIND, NULL, 0);
+    mkmount ("tmpfs", "dev/shm", "tmpfs", 0, NULL, 1);
+    mkmount ("/proc", "proc", "", MS_BIND | MS_REC, NULL, 1);
+    mkmount ("/sys", "sys", "", MS_BIND | MS_REC, NULL, 1);
+    mkmount ("tmpfs", "tmp", "tmpfs", 0, NULL, 1);
+#endif
+    if (chroot (".") != 0) {
+      fputs("Failed to chroot into .\n", stderr);
+      exit(EXIT_FAILURE);
+    }
+  }
+  free(cwd);
+
+
+  char **newenv = malloc(13 * sizeof(char *));
+  char **newenv_end = newenv;
+  if (newenv == NULL) {
+    fputs("Failed to allocate space for new environment\n", stderr);
+    exit(EXIT_FAILURE);
+  }
+
+  newenv_end = copy_environment(newenv_end, "ARCH");
+  newenv_end = copy_environment(newenv_end, "ARCH_DIR");
+  newenv_end = copy_environment(newenv_end, "M2LIBC");
+  newenv_end = copy_environment(newenv_end, "TOOLS");
+  newenv_end = copy_environment(newenv_end, "BLOOD_FLAG");
+  newenv_end = copy_environment(newenv_end, "BASE_ADDRESS");
+  newenv_end = copy_environment(newenv_end, "ENDIAN_FLAG");
+  newenv_end = copy_environment(newenv_end, "BINDIR");
+  newenv_end = copy_environment(newenv_end, "BUILDDIR");
+  newenv_end = copy_environment(newenv_end, "TMPDIR");
+  newenv_end = copy_environment(newenv_end, "OPERATING_SYSTEM");
+  newenv_end[0] = "WRAPPED=yes";
+  newenv_end[1] = NULL;
+
+
+#ifdef __M2__
+#if __uefi__
+  return spawn (argv[1], argv + sizeof(char *), newenv);
```

Comment:
Oh nice, you even made it transparent for UEFI! Just curious, have you tried running it there?

Reply:
Oh I guess you did, I can see your UEFI changes in another PR :).